File: internal/analyzers/text.go
package analyzers

import (
	"sort"
	"strings"

	"github.com/toddwbucy/Aletharsis/internal/evidence"
	u "github.com/toddwbucy/Aletharsis/internal/unicoderef"
)

type Text struct{}

func (Text) Analyze(d *evidence.Document) []evidence.Finding {
	result := []evidence.Finding{}
	for _, t := range d.Texts {
		active := 0
		for _, count := range t.LineEndings {
			if count > 0 {
				active++
			}
		}
		if active > 1 || t.LineEndings["cr"] > 0 {
			details := evidence.Object{}
			for k, v := range t.LineEndings {
				details[k] = v
			}
			result = append(result, finding("text.line_endings", details, evidence.Object{"source": t.Source}))
		}
		if t.BOM != nil {
			result = append(result, finding("text.bom", evidence.Object{"hex": *t.BOM, "encoding": t.Encoding}, evidence.Object{"source": t.Source, "byte_offset": 0}))
		}
		rs := []rune(t.Text)
		positions := []int{}
		for i, r := range rs {
			if i > 0 && r == 0xfeff {
				positions = append(positions, i)
			}
		}
		if len(positions) > 0 {
			result = append(result, finding("text.embedded_bom", evidence.Object{"count": len(positions)}, evidence.Location(t, positions)))
		}
		tokenRune := func(r rune) bool { return u.Word(r) && r != '_' && u.Category(r) != "Nd" }
		for start := 0; start < len(rs); {
			if !tokenRune(rs[start]) {
				start++
				continue
			}
			end := start + 1
			for end < len(rs) && tokenRune(rs[end]) {
				end++
			}
			scripts := map[string]bool{}
			for _, r := range rs[start:end] {
				if u.Letter(r) {
					scripts[strings.Split(u.Name(r, ""), " ")[0]] = true
				}
			}
			if scripts["LATIN"] && (scripts["GREEK"] || scripts["CYRILLIC"]) {
				keys := []string{}
				for k := range scripts {
					keys = append(keys, k)
				}
				sort.Strings(keys)
				result = append(result, finding("text.mixed_script", evidence.Object{"escaped_token": u.Escaped(string(rs[start:end])), "scripts": keys}, evidence.Location(t, []int{start})))
			}
			start = end
		}
	}
	return result
}

type Metadata struct{}

func (Metadata) Analyze(d *evidence.Document) []evidence.Finding {
	result := []evidence.Finding{}
	keys := []string{}
	for k := range d.Metadata {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		result = append(result, evidence.Finding{ID: "metadata.present", Severity: "INFO", Confidence: 1, Category: "metadata", Classification: "observed_fact", Title: k + " metadata present", Description: "Metadata is evidence, not proof of watermarking.", Evidence: evidence.Object{"key": k, "value": d.Metadata[k]}, Location: evidence.Object{}})
	}
	return result
}

File: src/aletharsis/analyzers/text.py
import re
import unicodedata as ud

from ..models import DocumentEvidence, Finding
from .unicode import escaped, location


class TextAnalyzer:
    def analyze(self, evidence: DocumentEvidence) -> list[Finding]:
        findings = []
        for segment in evidence.texts:
            active = [name for name, count in segment.line_endings.items() if count]
            if len(active) > 1 or "cr" in active:
                findings.append(Finding(
                    "text.line_endings", "LOW", 1.0, "document_structure", "observed_fact",
                    "Mixed or legacy line endings detected",
                    "Line-ending differences can result from ordinary editing or platform conversion.",
                    segment.line_endings, {"source": segment.source}))
            if segment.bom:
                findings.append(Finding("text.bom", "INFO", 1.0, "document_structure", "observed_fact",
                                        "Byte order mark detected", "An initial BOM is an ordinary encoding marker.",
                                        {"hex": segment.bom, "encoding": segment.encoding},
                                        {"source": segment.source, "byte_offset": 0}))
            embedded = [i for i, c in enumerate(segment.text) if c == "\ufeff" and i != 0]
            if embedded:
                findings.append(Finding("text.embedded_bom", "LOW", 1.0, "document_structure", "observed_fact",
                                        "Embedded BOM character detected", "Possible concatenated text or embedded formatting marker.",
                                        {"count": len(embedded)}, location(segment, embedded)))
            for match in re.finditer(r"[^\W\d_]+", segment.text):
                scripts = {ud.name(c, "").split(" ")[0] for c in match[0] if c.isalpha()}
                if "LATIN" in scripts and scripts & {"CYRILLIC", "GREEK"}:
                    findings.append(Finding(
                        "text.mixed_script", "LOW", 0.6, "unicode", "suspicious_pattern",
                        "Mixed Latin and Greek/Cyrillic letters within a token",
                        "Mixed scripts can contain visually similar letters, but can also be intentional. This is not a comprehensive confusables detector.",
                        {"escaped_token": escaped(match[0]), "scripts": sorted(scripts)},
                        location(segment, [match.start()])))
        return findings
